### ISSUE TITLE
Table layout will break if the content of a cell requires more than one page

### ISSUE THREAD

**Issue by SMMousaviSP** [py-pdf/fpdf2#1142]
When the content of one cell within a table is too long, the layout will break.
[table.pdf](https://github.com/py-pdf/fpdf2/files/14873843/table.pdf)

**Minimal code**
```python
from fpdf import FPDF

WEIRD_LONG_TEXT = "Just generate a very long text with any lorem ipsum generator, make sure it is long enough to require more than one page"

TABLE_DATA = (
    ("First name", "Last name", "Age", "City"),
    ("Jules", "Smith", "34", "San Juan"),
    ("Mary", "Ramos", "45", WEIRD_LONG_TEXT),
    ("Carlson", "Banks", "19", "Los Angeles"),
)

pdf = FPDF()
pdf.add_page()
pdf.set_font("Times", size=16)
with pdf.table() as table:
    for data_row in TABLE_DATA:
        row = table.row()
        for datum in data_row:
            row.cell(datum)
pdf.output('table.pdf')
```

**Environment**
Please provide the following information:
* **Operating System**: Fedora 39
* **Python version**: 3.11.8
* **`fpdf2` version used**: 2.7.8

**Comment by SMMousaviSP** [comment]
Hello @Lucas-C 
Thank you for your response.

What I had in mind was something like this. This is the default behavior of Microsoft Word.
[table_in_word.pdf](https://github.com/py-pdf/fpdf2/files/14885375/table_in_word.pdf)

I understand that if a text is too long, it may not make much sense to store it in a table. However, imagine a scenario where most of the time all cells of the table are going to be just one line or 2 to 3 lines and using the table would make sense and is a nice visual representation but once or twice a cell might become longer than expected. This will prevent the use of tables in automated tasks since we know it is possible for the table to break (even if it happens rarely).

As you saw in my first example, in the case of exceedingly long texts, it doesn't only break that specific row and the layout of the whole table gets broken.

**Comment by Lucas-C** [comment]
Hi @SMMousaviSP

I had a closer look at the code, and I think this will be a difficult addition.

Basically, it would mean to A) detect this situation and B) split the row into several smaller rows, which is something I am bit relunctant to do, because a consequence would be for the table final rows count to be unpredictable for the user...

A first step / easier solution could be to A) detect this situation and B) raise an error in that case.

What do you think about this strategy @SMMousaviSP?
Also pinging @mjasperse who recently worked on adding support for `rowspan` to tables.

**Comment by mjasperse** [comment]
I have some ideas for how to handle rendering a long cell across multiple pages, although I see the potential for a lot of edge cases where the desired behaviour is not necessarily clear. I'm happy to give it a go and see how far I get with a simplistic approach.

**Comment by SMMousaviSP** [comment]
@Lucas-C Thank you for taking time and looking into this issue.

Would there be a reasonable solution to resolve this error when it happens? For example by somehow figuring out how much of the text is fitting to be able to either cut the rest or make a new row for that manually? Otherwise to be honest I prefer to have the broken table compared to not having any PDF at all.

**Comment by Lucas-C** [comment]
> I have some ideas for how to handle rendering a long cell across multiple pages, although I see the potential for a lot of edge cases where the desired behaviour is not necessarily clear. I'm happy to give it a go and see how far I get with a simplistic approach.

You are very welcome to give it a try @mjasperse! 🙂
You can use PR #1143 as a starting point, or go your own way.

> Would there be a reasonable solution to resolve this error when it happens? For example by somehow figuring out how much of the text is fitting to be able to either cut the rest or make a new row for that manually?

Good idea @SMMousaviSP, dropping extraneous text seems easier to implement than splitting the cell/row.
A first naive implementation i think of would be to use a [dichotomic search](https://en.wikipedia.org/wiki/Dichotomic_search) in order to figure where in the text content to cut in order to maximize its length but still fit on the page.
If we adopt this approach, it would be nice to produce a log warning when such "cut" happens.

**Comment by mjasperse** [comment]
Since `multi_cell` already computes line breaks with word wrapping as part of the first layout pass, my idea was to use that to return the page break locations and recompute the table layout. I think it would be not so hard if there are no rowspans.

**Comment by AssiaSOntoforce** [comment]
Table with multi-page columns
http://www.fpdf.org/?go=script&id=12

**Comment by Lucas-C** [comment]
> Since `multi_cell` already computes line breaks with word wrapping as part of the first layout pass, my idea was to use that to return the page break locations and recompute the table layout. I think it would be not so hard if there are no rowspans.

OK!
How do you suggest that `fpdf2` should behave if rowspans are used then? A `NotImplementedError` would be raised?

**Comment by shoaib-moeen** [comment]
Hey @Lucas-C , Can I have a go at this one?

**Comment by mjasperse** [comment]
@shoaib-moeen you can give it a try! My simple idea broke rather quickly on the test suite because of scenarios like table cells with multiple "overflowing" cells, or images. If there is only one long cell, that is fairly easy to handle but I think the general-case complexity grows quickly.

Working on my initial implementation raised the following points that I would appreciate your thoughts on @Lucas-C 

1. When should a cell should be split versus when should the whole cell be moved to the next page? For example, if you had a cell that was 5 lines long where a pagebreak is anticipated after the first line, should it be broken? I think it would be better to move the whole cell to the next page in that case, so what is the cutoff? 3 lines? 5 lines? 50%? A user-specified parameter?
2. Should the user have the ability to specify a cell cannot be split? This is relevant for cells containing images, which should never be broken and that has implications for whether the adjacent cells are split or moved.
3. How should vertical alignment be handled? What does it mean to split a cell that is specified with vertical centering? Or a cell that now spans multiple pages because the adjacent cell is very large? Should as much as possible be put into the first page, or should it be centered in the "largest" part? I should check how Word handles this case.
4. It think cells should not be split if a rowspan is involved because it adds too much complexity. If the combined rowspan exceeds the EPH I think throwing an exception might be OK. Can be annoying to debug since it doesn't generate any PDF output in that scenario - maybe it makes sense to truncate the table and throw the exception such that catching and calling `pdf.output()` will make an output that can be checked?

Maybe I'm overthinking the edge cases and the primary use-case here is handling one long cell and ensuring a sensible PDF is generated at all. However the first two points affect the existing test-suite cases.

Unfortunately I doubt I will have time to work on this implementation in the next two months because of work commitments, sorry that I wasn't able to come up with something faster.

**Comment by Lucas-C** [comment]
> 1. When should a cell should be split versus when should the whole cell be moved to the next page? For example, if you had a cell that was 5 lines long where a pagebreak is anticipated after the first line, should it be broken? I think it would be better to move the whole cell to the next page in that case, so what is the cutoff? 3 lines? 5 lines? 50%? A user-specified parameter?

IMHO, for now there is no need for fine-grained control there, but we should leave it open for further improvements later on:
* for now, the behaviour could simply naive, that is, in the case you mentioned, the 1st line would be rendered on the first page, and the other lines on the 2nd page, and that's it
* the default behaviour should be the same as the current behaviour, _i.e._ NOT BREAK, to ensure backward compatibility
* we could introduce a new optional parameter passed to `Table` constructor, maybe named `break_cells` with a default value of `None` and a valid enum value of `IF_ROW_HEIGHT_REQUIRE_PAGE_BREAK`
* later on we will be able to add support more enum values, providing fine-grained control over cell-break logic

**Comment by gmischler** [comment]
> * When should a cell should be split versus when should the whole cell be moved to the next page? For example, if you had a cell that was 5 lines long where a pagebreak is anticipated after the first line, should it be broken? I think it would be better to move the whole cell to the next page in that case, so what is the cutoff? 3 lines? 5 lines? 50%? A user-specified parameter?

The typographical terms are "widows" (left-over line(s) after a page/column break) and "orphans" (left behind line(s) before a page/column break). High-end layout software usually allows to specify a minimum number for each (or no break if there are fewer lines than either limit).
This might become an issue for `text_columns()` as well, so we'll need a consistent solution in all relevant places. Given that table cells are also going to be converted into text regions, maybe we should postpone that question for after that has happened. In the end, it will probably rather apply to individual `paragraph()`s than columns or table cells.

> * Should the user have the ability to specify a cell cannot be split? This is relevant for cells containing images, which should never be broken and that has implications for whether the adjacent cells are split or moved.

This is an interesting idea, but will obviously only be enforcible if the cell contents fit on one page.
I don't think that images will really make any difference here though. The weird "image first, text after" rule currently in place will have to go away in text region based table cells anyway, since those will allow to add an arbitrary number of text chunks and images in any order.

> * How should vertical alignment be handled? What does it mean to split a cell that is specified with vertical centering? Or a cell that now spans multiple pages because the adjacent cell is very large? Should as much as possible be put into the first page, or should it be centered in the "largest" part? I should check how Word handles this case.

I would try to center content over the complete height of the column, no matter how many pages it spans. This makes the solution independent of whether the vertically centered content actually fits on any one of the involved pages. Top and bottom valigned content is obvious.

> * It think cells should not be split if a rowspan is involved because it adds too much complexity. If the combined rowspan exceeds the EPH I think throwing an exception might be OK. Can be annoying to debug since it doesn't generate any PDF output in that scenario - maybe it makes sense to truncate the table and throw the exception such that catching and calling `pdf.output()` will make an output that can be checked?

Let's cross that bridge when we get there... :wink:
I'm not sure if that really needs to be a problem, though.

You may notice that I keep referring to a future conversion to text regions. This will involve pretty much a rewrite of the base mechanics of how tables are built, since text regions use different concepts than the currently used multi_cells. The new code will be more stringently structured, delegate more functionality to lower levels, and thus hopefully simplify the overall architecture of the module. I had started working on this conversion several months ago. But then the table code repeatedly sprouted new features left and right, which made it difficult to keep track and adapt the rewrite accordingly.
I think at some point we'll have to establish a feature-freeze on the table code until I had a chance to chatch up.

**Comment by Lucas-C** [comment]
Thank you for your insight @gmischler 🙂

>  I think at some point we'll have to establish a feature-freeze on the table code until I had a chance to chatch up.

Sure, I agree with this idea, as long as it does not stall `fpdf2` development, and that we still welcome contributions from motivated developpers 😊

> You may notice that I keep referring to a future conversion to text regions. [...] I had started working on this conversion several months ago.

Maybe a next step regarding text regions could be to split the work and define clear GitHub issues that other contributors could pick up and work on?
This way the implementation of this feature would not only rely on your availability.

If I'm not mistaken, the initial design for text regions came this discussion 2 years ago: https://github.com/py-pdf/fpdf2/discussions/339
Maybe a new roadmap could be drafted based on the current state of things, in order to identify the steps required to get to the final set of features that you envision @gmischler?

**Comment by FitzKaos** [comment]
Hello, I came across this issue and I have got a bit of a hacky fix that seems to work for me - and apologies for the rough and ready nature, I just needed something to work and then I will focus on making it look pretty.

So I initially convert the required text to a pandas dataframe (in my case my source is json). Once that is done, I then wanted to find those cells that have too long strings and split them over multiple lines.
```
    for col in pd_df:
        for i, row_value in pd_df[col].items():
            if len(row_value) > 500:
                new_string = row_value[:500] + '/n' + row_value[500:]
                pd_df.loc[i, col] = new_string
        
        pd_df[col] = pd_df[col].str.split('/n')
        pd_df = pd_df.explode(col)

    df_tuple = pd_df.to_records(index=False).tolist()
    df_tuple.insert(0, tuple(pd_df.columns))
    with pdf.table() as table:
        for data_row in df_tuple:
            row = table.row()
            for datum in data_row:
                row.cell(str(datum))
```

This will split the long text into two parts, one that is 500 characters and then the remainder (n.b. 500 was just chosen as a nice round number). I still need to create a generalisable solution, since there are a number of issues with the above:
- I am aware that for _super_ long strings, the remainder might need splitting over multiple pages itself. 
- I chose `/n` as my break point since I know I will not see any others in my use-case - this might need to be changed for different use-cases. 
- Finally, this can (and does) split words in the middle, so a solution that would only split after a full word would be required.

However, this does work for my use-case and it might help provide inspiration for more capable programmers than I to solve the generalisable case.
